packages/engine/lib/error/src/macros.rs

```diff
@@ -0,0 +1,567 @@
+#[doc(hidden)]
+pub mod __private {
+    use core::fmt;
+
+    use crate::Report;
+
+    pub mod kinds {
+        use core::{fmt, marker::PhantomData};
+
+        use crate::Report;
+
+        pub trait AdhocKind: Sized {
+            fn __kind(&self) -> Adhoc {
+                Adhoc
+            }
+        }
+        impl<T> AdhocKind for &T where T: ?Sized + fmt::Display + fmt::Debug + Send + Sync + 'static {}
+
+        pub struct Adhoc;
+        impl Adhoc {
+            #[allow(clippy::unused_self)]
+            pub fn report<C>(self, context: C) -> Report
+            where
+                C: fmt::Display + fmt::Debug + Send + Sync + 'static,
+            {
+                Report::new(context)
+            }
+        }
+
+        pub trait TraitKind<S>: Sized {
+            fn __kind(&self) -> Trait<S> {
+                Trait(PhantomData)
+            }
+        }
+        impl<E, S> TraitKind<S> for E where E: Into<Report<S>> {}
+
+        pub struct Trait<S>(PhantomData<S>);
+        impl<S> Trait<S> {
+            #[allow(clippy::unused_self)]
+            pub fn report<E: Into<Report<S>>>(self, error: E) -> Report<S> {
+                error.into()
+            }
+        }
+    }
+
+    pub fn format_err(args: fmt::Arguments) -> Report {
+        Report::new(alloc::format!("{}", args))
+    }
+}
+
+/// Creates a [`Report`] from the given parameters.
+///
+/// Optionally a scope can be specified.
+///
+/// [`Report`]: crate::Report
+///
+/// # Examples
+///
+/// Create a [`Report`] from a message:
+///
+/// ```
+/// # fn has_permission(_: &User, _: &Resource) -> bool { false }
+/// # struct User;
+/// # struct Resource;
+/// # use core::fmt;
+/// # impl fmt::Display for Resource { fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result { Ok(()) }}
+/// use error::format_err;
+///
+/// # fn use_resource(user: User, resource: Resource) -> error::Result<()> {
+/// if !has_permission(&user, &resource) {
+///     return Err(format_err!("permission denied for accessing {resource}"));
```

Comment:
Do backtraces get generated when using this macro on a string?

Reply:
Yes, every `Report` has three location information, regardless on how they are generated.

Reply:
And do `Location`s get used if `Backtrace` is missing?

Reply:
Yes, the `Location` is the output like in the module level documentation below the error/causes. If `Backtrace` isn't captured, the part is just skipped from output and `Report::backtrace` returns `None`

Reply:
Right :) 
So all I'm asking is, are `Backtrace`s generated if we use this macro on a string?

Reply:
Yes, backtraces are generated (when `Report` is consructed) and are always guaranteed to be available on a `Report` (if enabled/compatible with your platform etc.).